ledger/complete/wal/checkpointer.go

```diff
@@ -430,7 +465,7 @@ func StoreCheckpoint(writer io.Writer, tries ...*trie.MTrie) error {
 		return fmt.Errorf("cannot write checkpoint footer: %w", err)
 	}
 
-	// Write CRC32 sum
+	// Write CRC32 sum for validation
```

Comment:
I wonder how to use this self-included checksum, because adding a checksum changes the checksum of the original file. Is there any tool that can automatically read the checksum from footer and compute the checksum of the remaining file and compare them?

Reply:
why do we need to write the checksum to the same file?  We can create the checkpoint file in a scratch directory, then atomic move the scratch directory to finalize the checkpoint.

Reply:
Found this which is close to what @pattyshack described, just not sure if there's a more idiomatic solution. https://en.wikipedia.org/wiki/Simple_file_verification#:~:text=Simple%20file%20verification%20(SFV)%20is,usually%20used%20for%20SFV%20files.

Reply:
The simplest solution is to maintain a manifest file, which keeps track of all the files that are part of this checkpoint and their corresponding checksum. (you will need to checksum the manifest file separately)

Another common approach is to divide the file into fixed size chunks, and checksum each chunk individually (e.g.,
https://github.com/google/leveldb/blob/main/doc/log_format.md)

Reply:
The current approach is to use a CRCWriter and CRCReader to generate and verify the checksum. 

When data bytes are passed to CRCWriter, it will compute the new checksum in memory, and then write the bytes to the file. When all bytes have been written to the file, it appends the checksum to the file. Since the checksum is appended at the end, when the CRCReader reads the bytes, it also computes the checksum, when it reaches the end before reading the checksum at the end, it will first compute the checksum and then compare with the checksum which is read by reading the file again. 

So the self-included checksum is mainly to be used by the CRCReader for validating, rather than external tool to validate.

Reply:
again, just because we've done something in certain ways in the past, doesn't mean we are stuck with them forever.

e.g., we can easily implement a different crc writer / reader that writes/reads the checksum to different file. i.e., suppose the crc writer is a io.WriteCloser, and we want to write file content into <blah>, the crc writer can create <blah>.crc32 when Close is called (and also closes <blah>).

Reply:
That's good point. I moved this idea into this issue for further discussion. 

https://github.com/dapperlabs/flow-go/issues/6402